Source/Plugins/bsfVulkanRenderAPI/MacOS/BsMacOSRenderWindow.mm

```diff
@@ -30,6 +30,83 @@
 		return getCore()->getCustomAttribute(name, pData);
 	}
 
+	void MacOSRenderWindow::initialize()
+	{
+		RenderWindowProperties& props = mProperties;
+		mIsChild = false;
+		
+		WINDOW_DESC windowDesc;
+		windowDesc.x = mDesc.left;
+		windowDesc.y = mDesc.top;
+		windowDesc.width = mDesc.videoMode.getWidth();
+		windowDesc.height = mDesc.videoMode.getHeight();
+		windowDesc.title = mDesc.title;
+		windowDesc.showDecorations = mDesc.showTitleBar;
+		windowDesc.allowResize = mDesc.allowResize;
+		windowDesc.modal = mDesc.modal;
+		windowDesc.floating = mDesc.toolWindow;
+		
+		auto iter = mDesc.platformSpecific.find("parentWindowHandle");
+		mIsChild = iter != mDesc.platformSpecific.end();
+		
+		mWindow = bs_new<CocoaWindow>(windowDesc);
+		mWindow->_setUserData(this);
+		
+		props.isFullScreen = mDesc.fullscreen && !mIsChild;
+		props.isHidden = mDesc.hidden;
+		
+		Rect2I area = mWindow->getArea();
+		props.width = area.width;
+		props.height = area.height;
+		props.top = area.y;
+		props.left = area.x;
+		props.hasFocus = true;
+		
+		props.hwGamma = mDesc.gamma;
+		props.multisampleCount = mDesc.multisampleCount;
+
+		// Create Vulkan surface
+		
+		// Add a CAMetalLayer to NSView
+		
+		NSView* view = (NSView*)mWindow->getView();
+		assert([view isKindOfClass:[NSView class]]);
+		
+		if (![view.layer isKindOfClass:[CAMetalLayer class]])
+		{
+			[view setLayer:[CAMetalLayer layer]];
+			[view setWantsLayer:YES];
+		}
+		
+		VkMacOSSurfaceCreateInfoMVK surfaceCreateInfo;
+		surfaceCreateInfo.sType = VK_STRUCTURE_TYPE_MACOS_SURFACE_CREATE_INFO_MVK;
+		surfaceCreateInfo.pNext = nullptr;
+		surfaceCreateInfo.flags = 0;
+		surfaceCreateInfo.pView = (__bridge const void *)(mWindow->getView());
+		
+		ct::VulkanRenderAPI& rapi = static_cast<ct::VulkanRenderAPI&>(ct::RenderAPI::instance());
```

Comment:
Hmm, I think this will work but it's very sketchy as it's relying on the core thread actually initializing Vulkan at this point. Which I think it should because there is a sync point elsewhere in code, but relying on some sync point elsewhere might not be the safest approach. How come this needs to be on the main thread and can be it moved to core thread?

Reply:
I know, but MoltenVK didn't allow to call vkCreateMacOSSurfaceMVK on non main thread.